packages/central-server/src/apiV2/surveyResponses/saveResponsesToDatabase.js

```diff
@@ -53,21 +53,29 @@ function buildResponseRecord(user, entitiesByCode, body) {
     id,
     entity_id: entityId,
     entity_code: entityCode,
-    timestamp,
     survey_id: surveyId,
     start_time: inputStartTime,
     end_time: inputEndTime,
     data_time: inputDataTime,
     approval_status: approvalStatus,
+    timestamp,
     timezone,
   } = body;
 
-  const time = new Date(timestamp).toISOString();
-  const startTime = inputStartTime ? new Date(inputStartTime).toISOString() : time;
-  const endTime = inputEndTime ? new Date(inputEndTime).toISOString() : time;
-  const dataTime = inputDataTime
-    ? stripTimezoneFromDate(new Date(inputDataTime).toISOString())
-    : stripTimezoneFromDate(time);
+  if (!timezone && !timestamp) {
+    throw new ValidationError(`Must provide timezone or timestamp`);
+  }
+
+  const defaultToTimestampOrThrow = (value, parameterName) => {
+    if (value) return new Date(value).toISOString();
+    if (timestamp) return new Date(timestamp).toISOString();
+
+    throw new ValidationError(`Must provide ${parameterName} or timestamp`);
+  };
+
+  const startTime = defaultToTimestampOrThrow(inputStartTime, 'start_time');
+  const endTime = defaultToTimestampOrThrow(inputEndTime, 'end_time');
+  const dataTime = stripTimezoneFromDate(defaultToTimestampOrThrow(inputDataTime, 'data_time'));
```

Comment:
Currently meditrak-app used the `/changes` route to submit data. That route does not require a `timestamp` field.
Once we switch to the meditrak-app-server, we will be calling the `/surveyResponse` route which does require a timestamp.

The timestamp is only needed if we are missing values for any of the following: `start_time`, `end_time`,`data_time`,`timezone`

This change will support a missing timestamp if all those values are present.

Reply:
Looks good, thanks for cleaning this up.

Happy as is, my only suggestion for making this a little bit clearer would be a big if-else, with the main logic first and then the fallback timestamp based logic after, with a comment that this is legacy.

Reply:
I'm not certain this is legacy or not. I think it's a valid for each of these fields to either provide them or rely on the `timestamp` field. It does seem a little odd though, doesn't it.

Reply:
Hmm odd, yeah either way a big ol comment there would be helpful :)